### PR TITLE
View refactor

### PR DESCRIPTION
**Pull request by cjmyers**
Substantial refactor of the views.  There are now only two view for SBOL:

collection.js
topLevelView.js

The customization is performed using the summarize<Type>.js methods in views/utils directory.  These were built off of sbolmeta, which has now been folded into SynBioHub.

This should not be merged in for a bit still as it needs some substantial testing.

### CODE REVIEW

File: lib/actions/createBenchlingSequence.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ var pug = require('pug')
 
 const { fetchSBOLObjectRecursive } = require('../fetch/fetch-sbol-object-recursive')
 
-var sbolmeta = require('sbolmeta')
+var sbolmeta = require('../views/utils/sbolmeta')
```
Comment (zachzundel) [inline lib/actions/createBenchlingSequence.js]: the refactored sbolmeta should just be a directory in lib

File: lib/bioschemas/DataRecord.js [hunk 2]
```diff
@@ -16,7 +16,7 @@ module.exports = function generateDataRecord(metadata) {
     "mainEntity": {
       "@type": metadata.rdfType,
       "identifier": metadata.name,
-      "hasBioChemRole": metadata.roles.map((role) => role.uri),
+      "hasBioChemRole": metadata.roles?metadata.roles.map((role) => role.uri):[],
```
Comment (zachzundel) [inline lib/bioschemas/DataRecord.js]: This operation is too complex for an in-line ternary operation, it is hard to read

File: lib/views/admin/theme.js [hunk 3]
```diff
@@ -70,8 +70,6 @@ function post(req, res) {
 
     localConfig.themeParameters[themeName] = newParameters 
 
-    config.set(localConfig)
```
Comment (zachzundel) [inline lib/views/admin/theme.js]: Why was this deleted?

File: lib/views/collection.js [hunk 4]
```diff
@@ -227,10 +185,16 @@ module.exports = function (req, res) {
             id: metaData.displayId,
             pid: metaData.persistentIdentity,
             wasDerivedFrom: metaData.wasDerivedFrom || '',
-            wasDerivedFroms: metaData.wasDerivedFrom?[ metaData.wasDerivedFrom ]:[],
+            wasDerivedFroms: metaData.wasDerivedFrom?
```
Comment (zachzundel) [inline lib/views/collection.js]: Too complicated for a ternary.

File: lib/views/collection.js [hunk 5]
```diff
@@ -227,10 +185,16 @@ module.exports = function (req, res) {
...
+		[ { uri: metaData.wasDerivedFrom,
+		    url: metaData.wasDerivedFrom } ]:[],
             wasGeneratedBy: { uri: metaData.wasGeneratedBy || '',
 			      url: uriToUrl(metaData.wasGeneratedBy,req)
 			    },
+            wasGeneratedBys: metaData.wasGeneratedBy?
```
Comment (zachzundel) [inline lib/views/collection.js]: Ternaries are hard to read.

File: lib/views/topLevelView.js [hunk 6]
```diff
@@ -0,0 +1,351 @@
+
+const { fetchSBOLObjectRecursive } = require('../fetch/fetch-sbol-object-recursive')
+const { getContainingCollections } = require('../query/local/collection')
+
```
Comment (zachzundel) [inline lib/views/topLevelView.js]: No need to put a space between imports

File: lib/views/topLevelView.js [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+		})
+            })
+
+        ])
+
+    }).then(function lookupEncodedProteins() {
+
+        var query =
+            'PREFIX sybio: <http://w3id.org/sybio/ont#>\n' +
+        'SELECT ?subject WHERE {' +
```
Comment (zachzundel) [inline lib/views/topLevelView.js]: fix indentation

File: lib/views/topLevelView.js [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+        return Promise.all([
+	    getContainingCollections(uri, graphUri, req.url).then((_collections) => {
+
+		collections = _collections
+
+		collections.forEach((collection) => {
+
+		    collection.url = uriToUrl(collection.uri)
+
+                    const collectionIcons = config.get('collectionIcons')
```
Comment (zachzundel) [inline lib/views/topLevelView.js]: Could move outside this function so that we do not read from disk/context switch for every collection

File: lib/views/topLevelView.js [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+                    
+                    if(collectionIcons[collection.uri])
+			collectionIcon = collectionIcons[collection.uri]
+		})
+            }),
+
+	    sparql.queryJson(getCitationsQuery, graphUri).then((results) => {
+
+                citations = results
```
Comment (zachzundel) [inline lib/views/topLevelView.js]: indentation

File: lib/views/topLevelView.js [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+            encodedProteins = results.map((result) => {
+                return result.subject
+            })
+
+        })
+
+    }).then(function getOtherComponentMetaData() {
+
+        if(meta.protein && meta.protein.encodedBy)
+            otherComponents = otherComponents.concat(meta.protein.encodedBy)
```
Comment (zachzundel) [inline lib/views/topLevelView.js]: if you still have todo here then wrap the if contents in {}

File: lib/views/topLevelView.js [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+		    }
+
+                })
+
+	    }
+        }
+
+    }).then(function fetchFromIgem() {
+
+	if(topLevel.wasDerivedFrom.toString().indexOf('http://parts.igem.org/') === 0) {
```
Comment (zachzundel) [inline lib/views/topLevelView.js]: Is this hitting the iGem registry on _every_ iGem part pageload? That's a lot, and will be slow. Could we maybe offload this onto a nightly job that updates from iGem?

Comment (zachzundel) [reply lib/views/topLevelView.js]: Or this should be a plugin.

File: lib/views/admin/theme.js (hunk 3 above)
Comment (cjmyers) [reply lib/views/admin/theme.js]: Not sure why.  Put it back.

File: lib/bioschemas/DataRecord.js (hunk 2 above)
Comment (cjmyers) [reply lib/bioschemas/DataRecord.js]: I think I prefer to keep this one.

File: lib/views/collection.js (hunk 4 above)
Comment (cjmyers) [reply lib/views/collection.js]: Agreed

File: lib/views/collection.js (hunk 5 above)
Comment (cjmyers) [reply lib/views/collection.js]: Agreed

File: lib/views/topLevelView.js (hunk 7 above)
Comment (cjmyers) [reply lib/views/topLevelView.js]: Fixed

File: lib/views/topLevelView.js (hunk 9 above)
Comment (cjmyers) [reply lib/views/topLevelView.js]: Seems right?

File: lib/views/topLevelView.js (hunk 10 above)
Comment (cjmyers) [reply lib/views/topLevelView.js]: Would like to move these things out of this file and wrap with only if CD.  Not sure how.

File: lib/views/topLevelView.js (hunk 11 above)
Comment (cjmyers) [reply lib/views/topLevelView.js]: Would be good to start by moving this out of the file into its own that is only hit on CD

**Comment by zachzundel** [comment]
Tried on some large files from the SBOL test suite, appears to be working

File: lib/attachments.js [hunk 12]
```diff
@@ -98,20 +99,33 @@ function getAttachmentsFromTopLevel(sbol, topLevel, share) {
...
-                type: attachment.format.toString().replace('http://identifiers.org/combine.specifications/','').replace('http://identifiers.org/edam/',''),
-                url: url,
-                size: attachment.size,
-                sizeString: attachment.size == -1 ? null : filesize(attachment.size)
-            })
-        }
+        if(attachment) {
+	    if (!(attachment instanceof URI)) {
+		var url = '/' + attachment.uri.toString().replace(config.get('databasePrefix'), '')
+		if (attachment.uri.toString().startsWith(config.get('databasePrefix') + 'user/') && share) {
```
[the remark this answers is not in the document]
Comment (cjmyers) [reply lib/attachments.js]: Perhaps.  The tricky thing is how to forward share links.  Namely, if you get a share link to a CD, we also give you share links to all its children and referenced Sequences.

File: lib/api/datatables.js [hunk 13]
```diff
@@ -73,28 +73,10 @@ function collectionMembersDatatable(req, res) {
...
-                    memberType = 'Sequence'
-                } else if (memberType === 'Model') {
-                    memberTypeUrl = 'http://wiki.synbiohub.org/wiki/Terms/SynBioHub#Model'
-                    memberType = 'Model'
-                } else if (memberType === 'Attachment') {
-                    memberTypeUrl = 'http://wiki.synbiohub.org/wiki/Terms/SynBioHub#Attachment'
-                    memberType = 'Attachment'
-                }
+		if (memberType === 'ComponentDefinition') memberType = 'Component'
+		else if (memberType === 'ModuleDefinition') memberType = 'Module'
```
[the remark this answers is not in the document]
Comment (cjmyers) [reply lib/api/datatables.js]: Yes, since these are all now handled uniformly.  Namely, the member type is stripped off the end of the type URI, and the URL is at http://sbolstandard.org now, since that is where Goksel's ontology is.
